File: screen-switcher/src/test/java/com/jaynewstrom/screenswitcher/AssertingScreenLifecycleListener.java
package com.jaynewstrom.screenswitcher;

import java.util.ArrayList;
import java.util.List;

final class AssertingScreenLifecycleListener implements ScreenLifecycleListener {

    private Screen activeScreen;
    private final List<Screen> screens = new ArrayList<>();

    @Override public void onScreenAdded(Screen screen) {
        if (!screens.add(screen)) {
            throw new AssertionError(screen + " screen already existed in: " + screens);
        }
    }

    @Override public void onScreenRemoved(Screen screen) {
        if (!screens.remove(screen)) {
            throw new AssertionError(screen + " screen did not exist in: " + screens);
        }
    }

    @Override public void onScreenBecameActive(Screen screen) {
        if (activeScreen != null) {
            throw new AssertionError(activeScreen + " was already active.");
        }
        if (!screens.contains(screen)) {
            throw new AssertionError(screen + " does not exist in screens: " + screens);
        }
        activeScreen = screen;
    }

    @Override public void onScreenBecameInactive(Screen screen) {
        if (activeScreen != screen) {
            throw new AssertionError("Active screen: " + activeScreen + " did not align with screen becoming inactive: " + screen);
        }
        if (!screens.contains(screen)) {
            throw new AssertionError(screen + " does not exist in screens: " + screens);
        }
        activeScreen = null;
    }
}

File: screen-switcher/src/test/java/com/jaynewstrom/screenswitcher/AssertingScreenLifecycleListener.kt
package com.jaynewstrom.screenswitcher

internal class AssertingScreenLifecycleListener : ScreenLifecycleListener {
    private var activeScreen: Screen? = null
    private val screens = ArrayList<Screen>()

    override fun onScreenAdded(screen: Screen) {
        if (!screens.add(screen)) {
            throw AssertionError(screen.toString() + " screen already existed in: " + screens)
        }
    }

    override fun onScreenRemoved(screen: Screen) {
        if (!screens.remove(screen)) {
            throw AssertionError(screen.toString() + " screen did not exist in: " + screens)
        }
    }

    override fun onScreenBecameActive(screen: Screen) {
        if (activeScreen != null) {
            throw AssertionError(activeScreen!!.toString() + " was already active.")
        }
        if (!screens.contains(screen)) {
            throw AssertionError(screen.toString() + " does not exist in screens: " + screens)
        }
        activeScreen = screen
    }

    override fun onScreenBecameInactive(screen: Screen) {
        if (activeScreen !== screen) {
            throw AssertionError("Active screen: $activeScreen did not align with screen becoming inactive: $screen")
        }
        if (!screens.contains(screen)) {
            throw AssertionError(screen.toString() + " does not exist in screens: " + screens)
        }
        activeScreen = null
    }
}
